tests/shared/utils/test_common.py

```diff
@@ -55,3 +56,65 @@ def test_transform_collection_to_sentence(
 ):
     actual = rasa.shared.utils.common.transform_collection_to_sentence(collection)
     assert actual in possible_outputs
+
+
+async def test_cached_method_with_sync_method():
+    expected = 5
+    mock = Mock(return_value=expected)
+
+    class Test:
+        @rasa.shared.utils.common.cached_method
+        def f(self):
+            return mock()
+
+    test_instance = Test()
+    assert test_instance.f() == expected
+    assert test_instance.f() == expected
+
+    mock.assert_called_once()
+
+
+async def test_cached_method_with_async_method():
+    expected = 5
+    mock = Mock(return_value=expected)
+
+    class Test:
+        @rasa.shared.utils.common.cached_method
+        async def f(self):
+            return mock()
+
+    test_instance = Test()
+    assert await test_instance.f() == expected
+    assert await test_instance.f() == expected
+
+    mock.assert_called_once()
```

Comment:
```suggestion
async def test_cached_method_with_async_method():
    expected = 5
    mock = Mock(return_value=expected)

    class Test:
        @rasa.shared.utils.common.cached_method
        async def f(self):
            return mock()

    test_instance = Test()
    assert await test_instance.f() == expected
    assert await test_instance.f() == expected

    mock.assert_called_once()


async def test_cached_method_with_async_method_and_sleep():
    expected = 5
    mock = Mock(return_value=expected)

    class Test:
        @rasa.shared.utils.common.cached_method
        async def f(self):
            await asyncio.sleep(1)
            return mock()

    test_instance = Test()
    assert await test_instance.f() == expected
    assert await test_instance.f() == expected

    mock.assert_called_once()
```

Reply:
Do we really need two test cases for that? Imo it's enough to add the `sleep` to the existing test function and be done with it. (I'd use `sleep(0)`  though)

Reply:
also fine with a single test :+1:

Reply:
Now I got what you meant with the test 👍